hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/DefaultS3ClientFactory.java

```diff
@@ -160,6 +182,84 @@ public AmazonS3 createS3Client(
     }
   }
 
+  /**
+   * Creates a new {@link S3Client}.
+   *
+   * @param uri S3A file system URI
+   * @param parameters parameter object
+   * @return S3 client
+   * @throws IOException on any IO problem
+   */
+  @Override
+  public S3Client createS3ClientV2(
+      final URI uri,
+      final S3ClientCreationParameters parameters) throws IOException {
+
+    Configuration conf = getConf();
+    bucket = uri.getHost();
+
+    final ClientOverrideConfiguration.Builder clientOverrideConfigBuilder =
+        AWSClientConfig.createClientConfigBuilder(conf);
+
+    final ApacheHttpClient.Builder httpClientBuilder =
+        AWSClientConfig.createHttpClientBuilder(conf);
+
+    final RetryPolicy.Builder retryPolicyBuilder = AWSClientConfig.createRetryPolicyBuilder(conf);
+
+    final ProxyConfiguration.Builder proxyConfigBuilder =
+        AWSClientConfig.createProxyConfigurationBuilder(conf, bucket);
+
+    S3ClientBuilder s3ClientBuilder = S3Client.builder();
+
+    // build a s3 client with region eu-west-2 that can be used to get the region of the bucket.
+    S3Client defaultS3Client = S3Client.builder().region(Region.EU_WEST_2)
+        .credentialsProvider(V1V2AwsCredentialProviderAdapter.adapt(parameters.getCredentialSet()))
+        .build();
+
+    // add any headers
+    parameters.getHeaders().forEach((h, v) -> clientOverrideConfigBuilder.putHeader(h, v));
+
+    if (parameters.isRequesterPays()) {
+      // All calls must acknowledge requester will pay via header.
+      clientOverrideConfigBuilder.putHeader(REQUESTER_PAYS_HEADER, REQUESTER_PAYS_HEADER_VALUE);
+    }
+
+    if (!StringUtils.isEmpty(parameters.getUserAgentSuffix())) {
+      clientOverrideConfigBuilder.putAdvancedOption(SdkAdvancedClientOption.USER_AGENT_SUFFIX,
+          parameters.getUserAgentSuffix());
+    }
+
+    clientOverrideConfigBuilder.retryPolicy(retryPolicyBuilder.build());
+    httpClientBuilder.proxyConfiguration(proxyConfigBuilder.build());
+
+    s3ClientBuilder.httpClientBuilder(httpClientBuilder)
+        .overrideConfiguration(clientOverrideConfigBuilder.build());
+
+    // use adapter classes so V1 credential providers continue to work. This will be moved to
+    // AWSCredentialProviderList.add() when that class is updated.
+    s3ClientBuilder.credentialsProvider(
+        V1V2AwsCredentialProviderAdapter.adapt(parameters.getCredentialSet()));
+
+    URI endpoint = getS3Endpoint(parameters.getEndpoint(), conf);
+
+    Region region =
+        getS3Region(conf.getTrimmed(AWS_REGION), defaultS3Client);
+
+    LOG.debug("Using endpoint {}; and region {}", endpoint, region);
+
+    s3ClientBuilder.endpointOverride(endpoint).region(region);
+
+    s3ClientBuilder.serviceConfiguration(
+        S3Configuration.builder().pathStyleAccessEnabled(parameters.isPathStyleAccess()).build());
```

Comment:
Move S3Configuration to temporary variable. There's a bunch of config in there that we may want to let users leverage in the future which could then just be one line changes.

```suggestion
    S3Configuration s3Configuration = S3Configuration.builder()
        .pathStyleAccessEnabled(parameters.isPathStyleAccess())
        .build();
    s3ClientBuilder.serviceConfiguration(s3Configuration);
```